lang/LangPrimSource/PyrUnixPrim.cpp

```diff
@@ -147,35 +152,29 @@ static void string_popen_thread_func(struct sc_process *process)
 int prString_POpen(struct VMGlobals *g, int numArgsPushed);
 int prString_POpen(struct VMGlobals *g, int numArgsPushed)
 {
-	struct sc_process *process;
 	PyrSlot *a = g->sp - 1;
 	PyrSlot *b = g->sp;
-	int err;
 
 	if (!isKindOfSlot(a, class_string)) return errWrongType;
 
-	char *cmdline = (char*)malloc(slotRawObject(a)->size + 1);
-	err = slotStrVal(a, cmdline, slotRawObject(a)->size + 1);
-	if(err) {
-		free(cmdline);
-		return errFailed;
-	}
+	char *cmdline = new char[slotRawObject(a)->size + 1];
+	slotStrVal(a, cmdline, slotRawObject(a)->size + 1);
 
 #ifdef SC_IPHONE
 	SetInt(a, 0);
 	return errNone;
 #endif
 
-	process = (struct sc_process *)malloc(sizeof(struct sc_process));
+	sc_process *process = new sc_process;
```

Comment:
Doing that:

```
diff --git a/lang/LangPrimSource/PyrUnixPrim.cpp b/lang/LangPrimSource/PyrUnixPrim.cpp
index a959d94..84b00d5 100644
--- a/lang/LangPrimSource/PyrUnixPrim.cpp
+++ b/lang/LangPrimSource/PyrUnixPrim.cpp
@@ -115,7 +115,7 @@ struct sc_process {
 	bool postOutput;
 };
 
-static void string_popen_thread_func(struct sc_process *process)
+static void string_popen_thread_func(std::unique_ptr<sc_process> process)
 {
 	FILE *stream = process->stream;
 	pid_t pid = process->pid;
@@ -134,7 +134,7 @@ static void string_popen_thread_func(struct sc_process *process)
 	if(process->postOutput)
 		postfl("RESULT = %d\n", res);
 
-	delete process;
+	//delete process;
 
 	gLangMutex.lock();
 	if(compiledOK) {
@@ -165,7 +165,7 @@ int prString_POpen(struct VMGlobals *g, int numArgsPushed)
 	return errNone;
 #endif
 
-	sc_process *process = new sc_process;
+	std::unique_ptr<sc_process> process(new sc_process);
 	process->stream = sc_popen(cmdline, &process->pid, "r");
 	setvbuf(process->stream, 0, _IONBF, 0);
 
@@ -174,7 +174,7 @@ int prString_POpen(struct VMGlobals *g, int numArgsPushed)
 	delete [] cmdline;
 
 	if(process->stream == NULL) {
-		delete process;
+		//delete process;
 		return errFailed;
 	}
 
@@ -235,14 +235,14 @@ int prArrayPOpen(struct VMGlobals *g, int numArgsPushed)
 		}
 	}
```

I get this error which is too advanced for me. Googling didn't help me much.

```
[ 34%] Building CXX object lang/CMakeFiles/libsclang.dir/LangPrimSource/PyrUnixPrim.cpp.o
In file included from /usr/include/c++/5/mutex:42:0,
                 from /usr/include/c++/5/condition_variable:39,
                 from /home/miguel/Development/SuperCollider/supercollider/common/SC_Lock.h:26,
                 from /home/miguel/Development/SuperCollider/supercollider/lang/LangSource/PyrSched.h:26,
                 from /home/miguel/Development/SuperCollider/supercollider/lang/LangPrimSource/PyrUnixPrim.cpp:33:
/usr/include/c++/5/functional: In instantiation of ‘struct std::_Bind_simple<std::_Bind<void (*(std::unique_ptr<sc_process>))(std::unique_ptr<sc_process>)>()>’:
/usr/include/c++/5/thread:137:59:   required from ‘std::thread::thread(_Callable&&, _Args&& ...) [with _Callable = std::_Bind<void (*(std::unique_ptr<sc_process>))(std::unique_ptr<sc_process>)>; _Args = {}]’
/home/miguel/Development/SuperCollider/supercollider/lang/LangPrimSource/PyrUnixPrim.cpp:181:60:   required from here
/usr/include/c++/5/functional:1505:61: error: no type named ‘type’ in ‘class std::result_of<std::_Bind<void (*(std::unique_ptr<sc_process>))(std::unique_ptr<sc_process>)>()>’
       typedef typename result_of<_Callable(_Args...)>::type result_type;
                                                             ^
/usr/include/c++/5/functional:1526:9: error: no type named ‘type’ in ‘class std::result_of<std::_Bind<void (*(std::unique_ptr<sc_process>))(std::unique_ptr<sc_process>)>()>’
         _M_invoke(_Index_tuple<_Indices...>)
         ^
In file included from /usr/include/c++/5/mutex:38:0,
                 from /usr/include/c++/5/condition_variable:39,
                 from /home/miguel/Development/SuperCollider/supercollider/common/SC_Lock.h:26,
                 from /home/miguel/Development/SuperCollider/supercollider/lang/LangSource/PyrSched.h:26,
                 from /home/miguel/Development/SuperCollider/supercollider/lang/LangPrimSource/PyrUnixPrim.cpp:33:
/usr/include/c++/5/tuple: In instantiation of ‘constexpr std::_Head_base<_Idx, _Head, false>::_Head_base(const _Head&) [with long unsigned int _Idx = 0ul; _Head = std::unique_ptr<sc_process>]’:
/usr/include/c++/5/tuple:357:21:   required from ‘constexpr std::_Tuple_impl<_Idx, _Head>::_Tuple_impl(const _Head&) [with long unsigned int _Idx = 0ul; _Head = std::unique_ptr<sc_process>]’
/usr/include/c++/5/tuple:473:33:   required from ‘constexpr std::tuple< <template-parameter-1-1> >::tuple(const _Elements& ...) [with _Elements = {std::unique_ptr<sc_process, std::default_delete<sc_process> >}]’
/usr/include/c++/5/functional:1114:70:   required from ‘std::_Bind<_Functor(_Bound_args ...)>::_Bind(_Functor&&, _Args&& ...) [with _Args = {std::unique_ptr<sc_process, std::default_delete<sc_process> >&}; _Functor = void (*)(std::unique_ptr<sc_process>); _Bound_args = {std::unique_ptr<sc_process, std::default_delete<sc_process> >}]’
/usr/include/c++/5/functional:1468:42:   required from ‘typename std::_Bind_helper<std::__is_socketlike<_Func>::value, _Func, _BoundArgs ...>::type std::bind(_Func&&, _BoundArgs&& ...) [with _Func = void (&)(std::unique_ptr<sc_process>); _BoundArgs = {std::unique_ptr<sc_process, std::default_delete<sc_process> >&}; typename std::_Bind_helper<std::__is_socketlike<_Func>::value, _Func, _BoundArgs ...>::type = std::_Bind<void (*(std::unique_ptr<sc_process>))(std::unique_ptr<sc_process>)>]’
/home/miguel/Development/SuperCollider/supercollider/lang/LangPrimSource/PyrUnixPrim.cpp:181:59:   required from here
/usr/include/c++/5/tuple:108:25: error: use of deleted function ‘std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = sc_process; _Dp = std::default_delete<sc_process>]’
       : _M_head_impl(__h) { }
                         ^
In file included from /usr/include/c++/5/condition_variable:43:0,
                 from /home/miguel/Development/SuperCollider/supercollider/common/SC_Lock.h:26,
                 from /home/miguel/Development/SuperCollider/supercollider/lang/LangSource/PyrSched.h:26,
                 from /home/miguel/Development/SuperCollider/supercollider/lang/LangPrimSource/PyrUnixPrim.cpp:33:
/usr/include/c++/5/bits/unique_ptr.h:356:7: note: declared here
       unique_ptr(const unique_ptr&) = delete;
       ^
```